eth/utils/bls.py

```diff
@@ -18,117 +21,163 @@
-        y *= HEX_ROOT
-    return y
-
-
-def hash_to_G2(m: bytes) -> Tuple[FQ2, FQ2, FQ2]:
-    """
-    WARNING: this function has not been standardized yet.
-    """
-    if m in CACHE:
-        return CACHE[m]
-    k2 = m
-    while 1:
-        k1 = blake(k2)
-        k2 = blake(k1)
-        x1 = int.from_bytes(k1, 'big') % field_modulus
-        x2 = int.from_bytes(k2, 'big') % field_modulus
-        x = FQ2([x1, x2])
-        xcb = x**3 + b2
-        if xcb ** ((field_modulus ** 2 - 1) // 2) == FQ2([1, 0]):
-            break
-    y = sqrt_fq2(xcb)
-    o = multiply((x, y, FQ2([1, 0])), 2 * field_modulus - curve_order)
-    CACHE[m] = o
-    return o
-
-
+#
+# G2
+#
 def compress_G2(pt: Tuple[FQ2, FQ2, FQ2]) -> Tuple[int, int]:
     assert is_on_curve(pt, b2)
     x, y = normalize(pt)
-    return (x.coeffs[0] + 2**255 * (y.coeffs[0] % 2), x.coeffs[1])
+    return (x.coeffs[0] + 2**383 * (y.coeffs[0] % 2), x.coeffs[1])
 
 
 def decompress_G2(p: bytes) -> Tuple[FQ2, FQ2, FQ2]:
-    x1 = p[0] % 2**255
-    y1_mod_2 = p[0] // 2**255
+    x1 = p[0] % 2**383
+    y1_mod_2 = p[0] // 2**383
     x2 = p[1]
     x = FQ2([x1, x2])
     if x == FQ2([0, 0]):
         return FQ2([1, 0]), FQ2([1, 0]), FQ2([0, 0])
-    y = sqrt_fq2(x**3 + b2)
+    y = modular_squareroot(x**3 + b2)
     if y.coeffs[0] % 2 != y1_mod_2:
         y = y * -1
-    assert is_on_curve((x, y, FQ2([1, 0])), b2)
+
     return x, y, FQ2([1, 0])
 
 
-def sign(m: bytes, k: int) -> Tuple[int, int]:
-    return compress_G2(multiply(hash_to_G2(m), k))
+#
+# APIs
+#
+def sign(message: bytes,
+         privkey: int,
+         domain: int) -> Tuple[int, int]:
+    return compress_G2(
+        multiply(
+            hash_to_G2(message, domain),
+            privkey
+        )
+    )
 
 
 def privtopub(k: int) -> int:
     return compress_G1(multiply(G1, k))
 
 
-def verify(m: bytes, pub: int, sig: bytes) -> bool:
+def verify(m: bytes, pub: int, sig: bytes, domain: int) -> bool:
     final_exponentiation = final_exponentiate(
         pairing(decompress_G2(sig), G1, False) *
-        pairing(hash_to_G2(m), neg(decompress_G1(pub)), False)
+        pairing(hash_to_G2(m, domain), neg(decompress_G1(pub)), False)
     )
     return final_exponentiation == FQ12.one()
 
 
-def aggregate_sigs(sigs: Iterable[bytes]) -> Tuple[int, int]:
+def aggregate_sigs(sigs: Sequence[bytes]) -> Tuple[int, int]:
     o = Z2
     for s in sigs:
         o = add(o, decompress_G2(s))
     return compress_G2(o)
 
 
-def aggregate_pubs(pubs: Iterable[int]) -> int:
+def aggregate_pubs(pubs: Sequence[int]) -> int:
     o = Z1
     for p in pubs:
         o = add(o, decompress_G1(p))
     return compress_G1(o)
+
+
+def verify_multiple(pubs: Sequence[int],
+                    msgs: Sequence[bytes],
+                    sig: bytes,
+                    domain: int) -> bool:
+    len_msgs = len(msgs)
+    assert len(pubs) == len_msgs
+
+    o = FQ12([1] + [0] * 11)
+    for m in set(msgs):
+        # aggregate the pubs
```

Comment:
I would gather the `pubs` related to `m` in a var maybe called `m_pubs` and then use `aggregate_pubs(m_pubs)` to get the `group_pub` rather than doing the individual calls to `add`.